File: Client/Assets/Common/Scripts/Net/Protocol/ObjectProxy.cs
#region Header
/**
 * 名称：NumberProxy
 
 * 日期：2015.11.16
 * 描述：
 **/
#endregion
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;

namespace NetCore
{
    public class ObjectProxy : Proxy
    {
        #region Fields
        
        #endregion


        #region Properties
    
        #endregion


        #region Constructors


        public ObjectProxy()
        {
    
        }
        #endregion


        #region Static Methods
    
        #endregion


        #region Private Methods
    
        #endregion

        public override bool getValue(IoBuffer ioBuffer, System.Type type, byte flag, out object value)
        {
            value = null;

            if (ioBuffer.ReadSize < 4)
            {
                Debuger.LogError("可读字节长度不足" + 4);
                return false;
            }

            int code = ioBuffer.ReadInt32();

            //取键数
            int fieldCount;
            if (!readVarInt32(ioBuffer, out fieldCount))
            {
                return false;
            }

            if (ioBuffer.ReadSize < 4)
            {
                Debuger.LogError("可读字节长度不足" + 4);
                return false;
            }

            var objLen = ioBuffer.ReadInt32();
            //减去objLen自己占用4字节
            objLen -= 4;

            if (ioBuffer.ReadSize < objLen)
            {
                Debuger.LogError("可读字节长度不足" + objLen);
                return false;
            }

            // 类型的获取
            TypeDef def = ProtocolCoder.instance.getTypeDef(code);
            if (def == null)
            {
                Debuger.LogError("类型定义{0}不存在", type.Name);
                return false;
            }
            if (def.type != type)
            {
                Debuger.LogError("类型不匹配。TypeDef:{0} Type:{1}", def.type.Name, type.Name);
                return false;
            }

            value = System.Activator.CreateInstance(type);

            //先保存读取位置，如果后面有多余的字段，就跳过去
            var readPos1 = ioBuffer.ReadPos;

            BeginParentLog(type.Name, fieldCount);
            var fieldCountInCfg = def.fields.Length;
            for (int i = 0; i < fieldCount; i++)
            {
                if (i < fieldCountInCfg)
                {
                    FieldInfo fieldInfo = def.fields[i];

                    AddLogNewLine(fieldInfo.Name + ": ");//加个分隔符
                    object obj;
                    if (!ProtocolCoder.instance.Decode(ioBuffer, fieldInfo.FieldType, out obj))
                    {
                        return false;
                    }
                    fieldInfo.SetValue(value, obj);
                    AddLog(",");//加个分隔符
                }
                else
                {
                    Debuger.Log("实际数据成员数比配置的多，主类型：{0}", type.Name);

                    //居然实现字段比描述的字段还多，看来对方加字段了，这里跳过这个对象剩余的数据吧
                    var readPos2 = ioBuffer.ReadPos;
                    //对象总长度减去已读的长度
                    ioBuffer.Skip(objLen - (readPos2 - readPos1));
                    break;
                }
            }
            EndParentLog();

            return true;
        }

        public override bool setValue(IoBuffer ioBuffer, object value)
        {
		    byte flag = Types.OBJECT;

            TypeDef def = ProtocolCoder.instance.getTypeDef(value.GetType());
            if (def == null)
            {
                if (ProtocolCoder.CanRegister(value.GetType()))
                    ProtocolCoder.instance.Register(value.GetType());
                def = ProtocolCoder.instance.getTypeDef(value.GetType());
                if (def == null)
                {
                    Debuger.LogError("找不到对象的预定义：{0}", value.GetType().Name);
                    return false;
                }
            }

            // 类型标记，#### 0000
            ioBuffer.Write(flag);

            //名字哈希值
            ioBuffer.Write(def.code);

            // 字段数量
            putVarInt32(ioBuffer, (int)def.fields.Length);

            //获取当前要写入的位置
            var writePos1 = ioBuffer.WritePos;
            //对象长度字段先占位
            ioBuffer.Write(0);

            BeginParentLog(value.GetType().Name, def.fields.Length);
            //Debuger.Log("{0} code:{1} 字段数:{2}", value, def.code, def.fields.Length);
            // 遍历属性
            for (int i = 0; i < def.fields.Length; i++)
            {
                AddLogNewLine(def.fields[i].Name + ": ");//加个分隔符
                if (!ProtocolCoder.instance.Encode(ioBuffer, def.fields[i].GetValue(value)))
                    return false;
                AddLog(",");//加个分隔符
            }
            EndParentLog();

            //再次获取当前要写入的位置
            var writePos2 = ioBuffer.WritePos;
            //两个位置相减得到对象的序列化大小
            var objectLen = writePos2 - writePos1;
            //更新序列化大小字段
            ioBuffer.WriteBack(objectLen, writePos1);

            return true;
        }

    

    }
}

File: Server/TestClient/app/libs/protocolProxy/objectProxy.js
"use strict";

var logUtil = require("../logUtil");
var BaseProxy = require("./baseProxy").BaseProxy;
var ProtocolCoder = require("../protocolCoder");

class ObjectProxy extends BaseProxy
{
    /**
     * 读取数据
     * @param {IOBuffer} ioBuf - 数据缓冲
     * @param {number} flag - 类型标记
     * @return {object} 读取的数据
     * @override
     */
    getValue(ioBuf, flag) {
        //var typeFlag = BaseProxy.getFlagType(flag);
        //if (typeFlag !== BaseProxy.TYPE_FLAG_OBJECT) {
        //    throw new Error("ObjectProxy~getValue 必须是Object类型");
        //}

        var coder = ProtocolCoder.instance;

        var code = ioBuf.readInt32();
        if (code === null) {
            throw new Error("ObjectProxy~getValue 数据字节不足" + 4);
        }

        //这里可能会抛出错误，让上层处理
        var fieldCount = BaseProxy.readVarUInt(ioBuf);

        var objLen = ioBuf.readInt32();
        if (objLen === null) {
            throw new Error("ObjectProxy~getValue 数据字节不足" + 4);
        }
        //减去objLen自己占用4字节
        objLen -= 4;

        if (ioBuf.canReadLen() < objLen) {
            throw new Error("ObjectProxy~getValue 数据字节不足" + objLen);
        }

        var objDesc = coder.getTypeDef(code);
        if (!objDesc) {
            throw new Error("ObjectProxy~getValue 未知类型HashCode：" + code);
        }

        //创建实例
        var obj = new objDesc.type();

        //先保存读取位置，如果后面有多余的字段，就跳过去
        var readPos1 = ioBuf.getReadPos();

        for (var i = 0; i < fieldCount; ++i) {
            var fieldDesc = objDesc.fields[i];
            if (fieldDesc) {
                var fieldVal = coder.decode(ioBuf);
                //如果是null，得看看允不允许null
                if (fieldVal === null || fieldVal === undefined) {
                    if (fieldDesc.notNull)
                        //Function.name暂时是非ES标准
                        throw new Error("ObjectProxy~getValue 成员不能为null，主类型：" + objDesc.name + "，成员名：" + fieldDesc.name);
                }
                //如果不是null，得用对类型
                else {
                    if (fieldVal.constructor !== fieldDesc.type)
                        //Function.name暂时是非ES标准
                        throw new Error("ObjectProxy~getValue 成员数据类型不对，主类型：" + objDesc.name + "，成员名：" + fieldDesc.name + "，实际类型：" + fieldVal.constructor.name + "，期望类型：" + fieldDesc.type.name);
                }

                obj[fieldDesc.name] = fieldVal;
            }
            else {
                logUtil.warn("ObjectProxy~getValue 实际数据成员数比配置的多，主类型：" + objDesc.name);

                //居然实现字段比描述的字段还多，看来对方加字段了，这里跳过这个对象剩余的数据吧
                var readPos2 = ioBuf.getReadPos();
                //对象总长度减去已读的长度
                ioBuf.skip(objLen - (readPos2 - readPos1));
                break;
            }
        }

        return obj;
    }

    /**
     * 写入数据
     * @param {IOBuffer} ioBuf - 数据缓冲
     * @param {object} value - 数据
     * @override
     */
    setValue(ioBuf, value)
    {
        //if (!Object.isObject(value))
        //{
        //    throw new Error("ObjectProxy~setValue 必须是Object");
        //}

        var coder = ProtocolCoder.instance;
        var type = value.constructor;
        if (!coder.canRegisterType(type))
        {
            throw new Error("ObjectProxy~setValue 类型必须带上字段描述");
        }

        var typeDef = coder.getTypeDef(type);
        if (!typeDef) {
            typeDef = coder.registerType(type);
            if (!typeDef) {
                throw new Error("ObjectProxy~setValue 注册失败：" + type);
            }
        }

        ioBuf.writeUInt8(BaseProxy.TYPE_FLAG_OBJECT);
        ioBuf.writeInt32(typeDef.code);

        var fields = typeDef.fields;
        var fieldCount = fields.length;
        BaseProxy.writeVarUInt(ioBuf, fieldCount);

        //获取当前要写入的位置
        var relWritePos1 = ioBuf.getRelativeWritePos();
        //对象长度字段先占位
        ioBuf.writeInt32(0);

        for (var i = 0; i < fieldCount; ++i) {
            var fieldDesc = fields[i];
            var fieldVal = value[fieldDesc.name];
            //如果是null，得看看允不允许null
            if (fieldVal === null || fieldVal === undefined) {
                if (fieldDesc.notNull)
                    //Function.name暂时是非ES标准
                    throw new Error("ObjectProxy~getValue 成员不能为null，主类型：" + type.name + "，成员名：" + fieldDesc.name);
            }
            //如果不是null，得用对类型
            else {
                if (fieldVal.constructor !== fieldDesc.type)
                    //Function.name暂时是非ES标准
                    throw new Error("ObjectProxy~getValue 成员数据类型不对，主类型：" + type.name + "，成员名：" + fieldDesc.name + "，实际类型：" + fieldVal.constructor.name + "，期望类型：" + fieldDesc.type.name);
            }

            coder.encode(ioBuf, fieldVal);
        }

        //再次获取当前要写入的位置
        var relWritePos2 = ioBuf.getRelativeWritePos();
        //两个位置相减得到对象的序列化大小
        var objectLen = relWritePos2 - relWritePos1;
        //更新序列化大小字段
        ioBuf.writeInt32WithRelativePos(objectLen, relWritePos1);
    }
}

exports.ObjectProxy = ObjectProxy;